packages/firestore/src/core/bundle.ts

```diff
@@ -98,3 +107,147 @@ export class BundleConverter {
+    totalBytes: metadata.totalBytes!
+  };
+}
+
+export class LoadResult {
+  constructor(
+    readonly progress: firestore.LoadBundleTaskProgress,
+    readonly changedDocs?: MaybeDocumentMap
+  ) {}
+}
+
+/**
+ * A class to process the elements from a bundle, load them into local
+ * storage and provide progress update while loading.
+ */
+export class BundleLoader {
+  /** The current progress of loading */
+  private progress: firestore.LoadBundleTaskProgress;
+  /**
+   * The threshold multiplier used to determine whether enough elements are
+   * batched to be loaded, and a progress update is needed.
+   *
+   * Applies to both `documentsBuffered` and `bytesBuffered`, triggers storage
+   * update and reports progress when either of them cross the threshold.
+   */
+  private thresholdMultiplier = 0.01;
+  /** Batched queries to be saved into storage */
+  private queries: bundleProto.NamedQuery[] = [];
+  /** Batched documents to be saved into storage */
+  private documents: BundledDocuments = [];
+  /**
+   * A BundleDocumentMetadata is added to the loader, it is saved here while
+   * we wait for the actual document.
+   */
+  private unpairedDocumentMetadata: bundleProto.BundledDocumentMetadata | null = null;
+
+  constructor(
+    private metadata: bundleProto.BundleMetadata,
+    private localStore: LocalStore
+  ) {
+    this.progress = initialProgress(metadata);
+  }
+
+  /**
+   * Adds an element from the bundle to the loader.
+   *
+   * If adding this element leads to actually saving the batched elements into
+   * storage, the returned promise will resolve to a `LoadResult`, otherwise
+   * it will resolve to null.
+   */
+  addSizedElement(
+    element: SizedBundleElement
+  ): firestore.LoadBundleTaskProgress | null {
+    debugAssert(!element.isBundleMetadata(), 'Unexpected bundle metadata.');
+
+    this.progress.bytesLoaded += element.byteLength;
+    if (element.payload.namedQuery) {
+      this.queries.push(element.payload.namedQuery);
+    }
+
+    if (element.payload.documentMetadata) {
+      if (element.payload.documentMetadata.exists) {
+        this.unpairedDocumentMetadata = element.payload.documentMetadata;
+      } else {
+        this.documents.push({
+          metadata: element.payload.documentMetadata,
+          document: undefined
+        });
+        this.progress.documentsLoaded += 1;
+      }
+    }
+
+    if (element.payload.document) {
+      debugAssert(
+        !!this.unpairedDocumentMetadata,
+        'Unexpected document when no pairing metadata is found'
+      );
+      this.documents.push({
+        metadata: this.unpairedDocumentMetadata!,
+        document: element.payload.document
+      });
+      this.progress.documentsLoaded += 1;
+      this.unpairedDocumentMetadata = null;
+    }
+
+    // Loading a document metadata will not update progress.
+    if (this.unpairedDocumentMetadata) {
+      return null;
+    }
+
+    return { ...this.progress };
```

Comment:
Note that I changed this in my diff to just retun this.progress, but I think you are correct and we need to return an immutable copy.